segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/ZKSegmentContainerMonitor.java

```diff
@@ -196,51 +202,63 @@ private void checkAssignment() {
-                        }
-                        try {
-                            // We remove the handle and don't attempt retry on stop container failures.
-                            unregisterHandle(containerId);
-                        } finally {
-                            // The pending task has to be removed after the handle is removed to avoid inconsistencies
-                            // with the container state.
-                            // Using finally block to ensure this is always called, otherwise this will prevent other
-                            // tasks from being attempted for this container id.
-                            this.pendingTasks.remove(containerId);
-                        }
-                    });
+            try {
+                return registry
+                        .stopContainer(handle, CLOSE_TIMEOUT_PER_CONTAINER)
+                        .whenComplete((aVoid, throwable) -> {
+                            if (throwable != null) {
+                                log.warn("Stopping container {} failed: {}", containerId, throwable);
+                            }
+                            try {
+                                // We remove the handle and don't attempt retry on stop container failures.
+                                unregisterHandle(containerId);
+                            } finally {
+                                // The pending task has to be removed after the handle is removed to avoid inconsistencies
+                                // with the container state.
+                                // Using finally block to ensure this is always called, otherwise this will prevent other
+                                // tasks from being attempted for this container id.
+                                this.pendingTasks.remove(containerId);
+                            }
+                        });
+            } catch (Throwable e) {
+                // The pending task has to be removed on all failures to enable retries.
+                this.pendingTasks.remove(containerId);
+                throw e;
+            }
         }
     }
 
     private CompletableFuture<ContainerHandle> startContainer(int containerId) {
         log.info("Starting Container {}.", containerId);
         this.pendingTasks.add(containerId);
-        return this.registry
-                .startContainer(containerId, INIT_TIMEOUT_PER_CONTAINER)
-                .whenComplete((handle, ex) -> {
-                    try {
-                        if (ex == null) {
-                            if (this.handles.putIfAbsent(handle.getContainerId(), handle) != null) {
-                                log.warn("Starting container {} succeeded but handle is already registered.",
-                                        handle.getContainerId());
+        try {
+            return this.registry
+                    .startContainer(containerId, INIT_TIMEOUT_PER_CONTAINER)
```

Comment:
The monitor thread runs at regular intervals to ensure container start/stop state is correctly maintained (hence failures are retried).